R/recycle.R

```diff
@@ -1,31 +1,125 @@
-#' Vector recycling
+#' Recycling rules used by r-lib and the tidyverse
+#'
+#' @description
+#' Recycling describes the concept of repeating elements of one vector to match
+#' the size of another vector. There are two rules that guide the recycling
+#' process used by packages in r-lib and the tidyverse:
+#'
+#' - Vectors of size 1 will be recycled to the size of any other vector
+#'
+#' - Otherwise, all vectors must have the same size
+#'
+#' @section Examples:
+#'
+#' Vectors of size 1 are recycled to the size of any other vector:
+#'
+#' ```
+#' vec_recycle_common(1:3, 1)
+#' #> [[1]]
+#' #> [1] 1 2 3
+#' #>
+#' #> [[2]]
+#' #> [1] 1 1 1
+#' ```
+#'
+#' This includes vectors of size 0:
+#'
+#' ```
+#' vec_recycle_common(integer(), 1)
+#' #> [[1]]
+#' #> integer(0)
+#' #>
+#' #> [[2]]
+#' #> numeric(0)
+#' ```
+#'
+#' If vectors aren't size 1, they must all be the same size:
+#'
+#' ```
+#' # This is fine, all vectors are the same size
+#' vec_recycle_common(1:3, 4:6)
+#' #> [[1]]
+#' #> [1] 1 2 3
+#' #>
+#' #> [[2]]
+#' #> [1] 4 5 6
 #'
-#' `vec_recycle(x, size)` recycles a single vector to given size.
-#' `vec_recycle_common(...)` recycles multiple vectors to their common size.
-#' All functions obey the vctrs recycling rules, described below, and will
-#' throw an error if recycling is not possible. See [vec_size()] for the
-#' precise definition of size.
+#' # This is an error
+#' vec_recycle_common(1:3, 4:7)
+#' #> Error:
+#' #> ! Can't recycle `..1` (size 3) to match `..2` (size 4).
+#' ```
 #'
-#' @section Recycling rules:
-#' The common size of two vectors defines the recycling rules, and can be
-#' summarise with the following table:
+#' @section Differences with base R:
```

Comment:
There is an appendix section in this vignette that talks about this too. I drew some examples from there, but again left it unchanged https://vctrs.r-lib.org/articles/type-size.html#appendix-recycling-in-base-r